sqlglot/optimizer/eliminate_subqueries.py

```diff
@@ -1,48 +1,153 @@
-    queries = {}
+    root = build_scope(expression)
 
-    for scope in traverse_scope(expression):
-        query = scope.expression
-        queries[query] = queries.get(query, []) + [query]
+    # Map of alias->Scope|Table
+    # These are all aliases that are already used in the expression.
+    # We don't want to create new CTEs that conflict with these names.
+    taken = {}
 
-    sequence = itertools.count()
+    for scope in root.traverse():
+        taken.update(
+            {
+                alias: source
+                for alias, source in scope.sources.items()
+                # Don't include derived tables that we're going to convert to CTEs,
+                # since these names won't be taken for long...
+                if not isinstance(source, Scope) or not source.is_derived_table
+            }
+        )
 
-    for query, duplicates in queries.items():
-        if len(duplicates) == 1:
-            continue
+    # Map of Expression->alias
+    # Existing CTES in the root expression. We'll use this for deduplication.
+    existing_ctes = {}
 
-        alias_ = f"_e_{next(sequence)}"
+    with_ = root.expression.args.get("with")
+    if with_:
+        for cte in with_.expressions:
+            existing_ctes[cte.this] = cte.alias
+    new_ctes = []
 
-        for dup in duplicates:
-            parent = dup.parent
-            if isinstance(parent, exp.Subquery):
-                parent.replace(alias(table(alias_), parent.alias_or_name, table=True))
-            elif isinstance(parent, exp.Union):
-                dup.replace(select("*").from_(alias_))
+    # We're adding more CTEs, but we want to maintain the DAG order.
+    # Derived tables within an existing CTE need to come before the existing CTE.
+    for cte_scope in root.cte_scopes:
+        # Append all the new CTEs from this existing CTE
+        for scope in cte_scope.traverse():
+            new_cte = _eliminate(scope, existing_ctes, taken)
+            if new_cte:
+                new_ctes.append(new_cte)
 
-        expression.with_(alias_, as_=query, copy=False)
+        # Append the existing CTE itself
+        new_ctes.append(cte_scope.expression.parent)
+
+    # Now append the rest
+    for scope in itertools.chain(root.union_scopes, root.subquery_scopes, root.derived_table_scopes):
+        for child_scope in scope.traverse():
+            new_cte = _eliminate(child_scope, existing_ctes, taken)
+            if new_cte:
+                new_ctes.append(new_cte)
+
+    if new_ctes:
+        expression.set("with", exp.With(expressions=new_ctes))
 
     return expression
+
+
+def _eliminate(scope, existing_ctes, taken):
+    if scope.is_union:
+        return _eliminate_union(scope, existing_ctes, taken)
+
+    if (scope.is_derived_table and not isinstance(scope.expression, (exp.Unnest, exp.Lateral))) or (
+        scope.is_subquery
+        and not scope.is_correlated_subquery
+        and not scope.expression.find_ancestor(exp.Having, exp.Join)
```

Comment:
why do you check if the ancestor is a join?

Reply:
duckdb doesn't like it:
```sql
WITH b AS (
  SELECT 1
)
SELECT
  *
FROM x
GROUP BY a
HAVING a > b
# Binder Error: column b must appear in the GROUP BY clause or be used in an aggregate function
```

```sql
WITH b AS (
  SELECT 1
)
SELECT
  *
FROM x
JOIN y
  ON y.a = b
# RuntimeError: Binder Error: Referenced column "b" not found in FROM clause!
# Candidate bindings: "y.a", "x.a"
# LINE 9:   ON y.a = b
#                    ^
```

The tpc-h tests caught this.

This bug existed before in eliminate_subqueries - we just weren't catching it because there were never any common subqueries in an ON or HAVING clause.